libqpdf/QPDF_Stream.cc

```diff
@@ -216,29 +216,27 @@ QPDF_Stream::getStreamJSON(
     auto dict = this->stream_dict;
     JSON result = JSON::makeDictionary();
     if (json_data != qpdf_sj_none) {
-        std::shared_ptr<Buffer> buf;
+        Pl_Discard discard;
+        Pl_Buffer buf_pl{"stream data"};
+        bool buf_pl_valid = false;
         bool filtered = false;
         bool filter = (decode_level != qpdf_dl_none);
         for (int attempt = 1; attempt <= 2; ++attempt) {
-            Pl_Discard discard;
-            std::shared_ptr<Pl_Buffer> buf_pl;
-            Pipeline* data_pipeline = nullptr;
+            Pipeline* data_pipeline = &discard;
             if (json_data == qpdf_sj_file) {
                 // We need to capture the data to write
-                buf_pl = std::make_shared<Pl_Buffer>("stream data");
-                data_pipeline = buf_pl.get();
-            } else {
-                data_pipeline = &discard;
+                data_pipeline = &buf_pl;
             }
             bool succeeded =
                 pipeStreamData(data_pipeline, &filtered, 0, decode_level, false, (attempt == 1));
-            if ((!succeeded) || (filter && (!filtered))) {
+            if (!succeeded || (filter && !filtered)) {
                 // Try again
                 filter = false;
                 decode_level = qpdf_dl_none;
+                buf_pl.getString(); // reset buf_pl
             } else {
-                if (buf_pl.get()) {
-                    buf = buf_pl->getBufferSharedPointer();
+                if (json_data == qpdf_sj_file) {
+                    buf_pl_valid = true;
```

Comment:
This is less safe than the original code. I think it's correct though. Now we're setting buf_pl_valid based on knowing here that we initialize it based on the same condition earlier. Before, it was setting it based on actually checking the validity of the pointer. The new code is less defensive and would be more likely to result in a crash if the code were changed. I think this is a difference in our coding styles.

Reply:
I changed buff_pl to a Pl_Buff rather than a shared_ptr<Pl_Buff>, so the object itself will always be valid. Perhaps rename to buf_pl_ready to express that buf_pl.finish() has been called and valid data can be retrieved?